### ISSUE TITLE
Preparing & uploading ITS_LIVE data

### ISSUE THREAD

**Issue by jhkennedy** [ASFHyP3/OpenData#10]
In order to cause the least amount of disruption to users, we'd like to migrate the `s3://its-live-data` bucket from the JPL ITS_LIVE account to the new OpenData - ITS_LIVE account. AWS bucket names are *globally unique* and there is currently no way in AWS to move buckets across accounts or even rename buckets. So, this will entail:
1. creating a temporary bucket
2. Copying all data from `s3://its-live-data` (JPL) to the temporary bucket
3. **verifying** all data has been moved to the temporary bucket
4. emptying the `s3://its-live-data` (JPL)
5. deleting `s3://its-live-data` (JPL) and *rapidly* creating `s3://its-live-data` (ASF) to reclaim the name
6. Copying all data from the temporary bucket to `s3://its-live-data` (ASF)
7. **verifying** all data has been moved to `s3://its-live-data` (ASF)
8. emptying and deleting the temporary bucket

However, this will be quite the task as there is ~150 TB of data and >1 billion objects in `s3://its-live-data` (JPL)
![image](https://github.com/ASFHyP3/OpenData/assets/7882693/00ca16fe-82d3-4ea4-beca-dc34a9766434)

The current options for migrating the data are:
a. AWS CLI via `aws s3 sync`
b. AWS CLI via individual copy commands (e.g., `aws s3 cp [SOURCE] [TARGET] `a billion times, likely split into distinct batches running in parallel.
c. [AWS S3 Batch Operations](https://aws.amazon.com/s3/features/batch-operations/)
d. [AWS S3 Batch Replication](https://docs.aws.amazon.com/AmazonS3/latest/userguide/s3-batch-replication-batch.html), which uses Batch Operations on the back end
e. Write a lamba/application to do it and fire it a billion times (see "Use an AWS SDK" in [this article](https://repost.aws/knowledge-center/s3-large-transfer-between-buckets))
f. [Use S3DistCp with Amazon EMR](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/UsingEMR_s3distcp.html)
g. [Use AWS DataSync](https://aws.amazon.com/datasync/)

Importantly, there will be both a **time** and **monetary** cost for these options, which should be estimated for all of them. We'll likely hit service limits no matter what we do, and after some back-of-the-envelope calculations, we could be looking at **3 weeks** to transfer all the data. 

**Alternatively**: we may want to consider using a new bucket name for AWS OpenData, and should evaluate the impact to users.

---

TODOs:
- [x] Remove project (non-user facing) data from `s3://its-live-open`
- [x] #15
- [x] Final transfer and verification

**Comment by jhkennedy** [comment]
Notably, we can use the HAND data (#9) to test out one of these options.

**Comment by jhkennedy** [comment]
@alex-s-gardner, @markf6, @mliukis, @betolink, how much would it suck (for us, and esp. users) to change the bucket name?

**Comment by jhkennedy** [comment]
We discussed Cutting over to a new bucket name in the ITS_LIVE project meeting on 2023-12-07.

*If* we go with a new bucket name, metadata will also need to be updated in a lot of places
* datacubes
* autorift velocity image pairs
* geojson/stac catalogs
* some code/tools
* some others I'm probably forgetting

Due to these changes, instead of transferring some of the data, it might make sense to regenerate the datacubes to fix all the metadata, but that's ~*a month* of effort
  * this might be a good proxy overall for how long it will take to update all the metadata

When I asked, "What's the data outage threshold for a new bucket instead?" the initial gut reaction was 2 weeks, but when we discussed all of the above updates we'll need to make, the threshold kept growing.

Overall, it's probably easiest to move the data twice, triaging `s3://its-live-data` into 3 categories:
1. project data -- should me moved to a project bucket @ JPL
2. user data -- should be moved to AWS OpenData
3. deprecated data -- should be deleted

I'll begin working on triaging the data now.

**Comment by jhkennedy** [comment]
Inventory of the top-level prefixes:

(1) project data -- should be moved to `s3://its-live-project`
 - [x] :question: `documentation/`
 - [x] :question: `ice_masks/`
 - [x] :question: `ice_shelf/`
 - [x] :question: `qgis_project/`
 - [x] `test_datacubes/`
   - [x] `test_datacubes/forAlex/`
   - [x] `test_datacubes/mosaics/`
   - [x] `test_datacubes/s1_correction/`
   - [x] `test_datacubes/validate_v2_granule_crop/`
 - [x] `velocity_image_pair/landsatOLI-latest/`
 - [x] `velocity_image_pair/sentinel1-backup/`
 - [x] `velocity_image_pair/sentinel1-corrected-8granules/`
 - [x] `velocity_image_pair/sentinel1-corrected/`
 - [x] `velocity_image_pair/sentinel1-latest/`
 - [x] `velocity_image_pair/sentinel2-latest/`
 - [x] :question: `elevation/`
- [x] :question: `height_change/`
 - [x] `month-data-logs/` --  monthly logs from 2022

(2) user data -- should be moved to AWS OpenData
 - [x] `autorift_parameters/` -- parameter files used to produce velocity image pairs
 - [x] `catalog_geojson/` -- GeoJSON catalog of all velocity image pairs
 - [x] `composites/`
 - [x] `datacubes/`
 - [x] `mosaics/`
 - [x] `rgb_mosaics/`
 - [x] `vel_web_tiles/`
 - [x] `velocity_image_pair/`
   - [x] `velocity_image_pair/landsatOLI/`
   - [x] `velocity_image_pair/sentinel1/`
   - [x] `velocity_image_pair/sentinel2/`
 - [x] `velocity_mosaic/`

(3) deprecated data -- should be ~~deleted~~ moved to `s3://its-live-project`
 - [x] :question: `L7_PV_fix/` -- ?
 - [x] :question: `Test/` -- ?
 - [x] :question: `isce_autoRIFT/` -- looks like old parameter files (circa 2021)
 - [x] :question: `test/` -- ?

(4) Does not need to be transferred:
- `NSIDC/` -- data archived at NSIDC
- `inventory/` -- AWS S3 inventory report we generated for this move

(5) Move from `its-live-project` to `its-live-open`:
- [x] `height_change`
- [x] `documentation`
- [x] `ice_masks`
- [x] `qgis_project`

(6) Misc:
- [x] `ice_shelf` can be deleted from `its-live-project`
   
~~@mliukis, can you look this over? Did I get everything in the right categories? In particular, I'm unsure of ones I've marked with a :question:.~~

**Comment by mliukis** [comment]
@jhkennedy I think @alex-s-gardner is the best candidate to confirm questionable ones

`s3://its-live-data/elevation` is just a Zarr format of the elevation data that Luis was using by some of the web apps to speed up the loading of the data, I believe. Not sure if it is the **user data**

:heavy_check_mark: `s3://its-live-data/velocity_image_pair` contains some of the original and corrected granules. User data should only include:
* /landsatOLI
* /sentinel1
* /sentinel2

I am keeping original S1 and intermittent corrected S1 granules for now as we need to fix encoding attributes for them - one of our high priority items once we are done with mosaics

**Comment by jhkennedy** [comment]
We're going to go ahead and sync all (1) project data and (3) deprecated data listed above to the new project bucket `s3://its-live-project` *EXCEPT*:
* `test_datacubes/` pending a better inventory of what's needed as it's about 1/3 of the total objects
* `velocity_mosaics/` pending confirmation that we want to keep it or not
* `month-data-logs/` pending confirmation that we want to keep it or not
* `inventory/` as it's no longer needed
* `NSIDC/` as it's all been archived at NSIDC

**Comment by mliukis** [comment]
@jhkennedy We should **keep** only the following sub-folders of the `test_datacubes`:
* forAlex
* mosaics
* s1_correction
* validate_v2_granule_crop

**Comment by jtherrmann** [comment]
The checklist in https://github.com/ASFHyP3/OpenData/issues/10#issuecomment-1850890480 (above) is getting a bit complex, so I re-grouped the prefixes into a list of what we expect to see in the `its-live-project` and `its-live-open` buckets and posted the results in https://github.com/ASFHyP3/OpenData/issues/15#issuecomment-1867140759.

**Comment by jtherrmann** [comment]
For the final transfer, I'm thinking we'll want to handle the prefixes in these groups:

**1. Run a sync command for each of these prefixes:**

```
s3://its-live-open/datacubes/definition/
s3://its-live-open/documentation/
s3://its-live-open/qgis_project/
```

**2. Submit a single Batch job for each of these prefixes:**

14 prefixes:

```
s3://its-live-open/autorift_parameters/
s3://its-live-open/catalog_geojson/
s3://its-live-open/height_change/
s3://its-live-open/mosaics/
s3://its-live-open/rgb_mosaics/
s3://its-live-open/vel_web_tiles/
s3://its-live-open/velocity_mosaic/
s3://its-live-open/ice_masks/
s3://its-live-open/composites/annual/v02_bad_v_error/
s3://its-live-open/composites/annual/v02_latest/
s3://its-live-open/composites/annual/v02_old_phase_amp/
s3://its-live-open/composites/annual/v02/
s3://its-live-open/composites/annual/v2/
s3://its-live-open/velocity_image_pair/sentinel1/v02/
```

**3. Submit Batch jobs for sub-prefixes of these prefixes:**

Submit a single Batch job for each sub-prefix under these parent prefixes. The following three prefixes contain 220, 192, and 173 sub-prefixes, respectively:

```
s3://its-live-open/datacubes/v2/
s3://its-live-open/velocity_image_pair/landsatOLI/v02/
s3://its-live-open/velocity_image_pair/sentinel2/v02/
```

**4. Copy these objects:**

```
s3://its-live-open/datacubes/catalog_v02.json
s3://its-live-open/velocity_image_pair/sentinel2/v02/used_granules.json
```

**Comment by jtherrmann** [comment]
**Steps for performing the final transfer:**

0. Run this command from the `batch-transfer/` directory to sync one of the smaller prefixes, and confirm that the job succeeds:
    ```
    make submit-job src_bucket=its-live-open dst_bucket=its-live-data s3_prefix=documentation/ job_name=test-for-final-transfer profile=its-live
    ```
1. Run these commands to sync the smaller prefixes:
    ```
    aws --profile opendata-its-live s3 sync s3://its-live-open/datacubes/definition/ s3://its-live-data/datacubes/definition/
    ```
    ```
    aws --profile opendata-its-live s3 sync s3://its-live-open/qgis_project/ s3://its-live-data/qgis_project/
    ```
2. Run the latest version of https://github.com/ASFHyP3/OpenData/blob/batch-transfer/batch-transfer/final_transfer.py (first with no arguments to perform a dry-run, then with `--submit` to submit jobs for real)
    * The output should show the script submitting 599 Batch jobs, one for each prefix (14+220+192+173 = 599, see the prefix counts in the previous comment)
4. Use `aws s3 cp` to copy these objects:
    ```
    s3://its-live-open/datacubes/catalog_v02.json
    s3://its-live-open/velocity_image_pair/sentinel2/v02/used_granules.json
    ```

**Comment by jhkennedy** [comment]
its-live-open -> its-live-data transfer wrapped up Sat. afternoon Alaska time, so we're now starting in on verifying the bucket contents before removing the temporary its-live-open bucket. First up, bucket metrics in the AWS console.

It looks like the gross total size of both buckets is the same, but has different storage tiers (?), while the total number of objects is the same:

* Its-live-data:
  * 71.4 TB Standard; 17.9 TB ITFA -- 89.3 TB total
  * 549,486,578 objects
* its-live-open:
  * 10.3 TB Standard; 27.1 MB ITIA; 79 TB ITFA -- 89.3 TB Total
  * 549,486,578 objects

**Comment by jtherrmann** [comment]
I ran https://github.com/ASFHyP3/OpenData/blob/batch-transfer/batch-transfer/check_sizes.py to compute total number of keys and total size per top-level prefix, and got the exact same results for the `its-live-open` and `its-live-data` inventory reports:

```
{'autorift_parameters': {'keys': 7806, 'size': 200130971664},
 'catalog_geojson': {'keys': 1412, 'size': 54820335094},
 'composites': {'keys': 3053872, 'size': 1224737415510},
 'datacubes': {'keys': 480065254, 'size': 46531848125930},
 'documentation': {'keys': 18, 'size': 74030080},
 'height_change': {'keys': 17, 'size': 41282341212},
 'ice_masks': {'keys': 1, 'size': 2625222151},
 'mosaics': {'keys': 412, 'size': 306036709833},
 'qgis_project': {'keys': 9, 'size': 93790},
 'rgb_mosaics': {'keys': 826, 'size': 160833084704},
 'vel_web_tiles': {'keys': 373533, 'size': 2862829097},
 'velocity_image_pair': {'keys': 65971064, 'size': 49137302630845},
 'velocity_mosaic': {'keys': 12354, 'size': 468987538103}}
```

Edit: Total number of keys = 549,486,578. Total size = `98131541328013` ~= 89.25 TB. These values match what @jhkennedy posted above. The size also matches what I had posted at https://github.com/ASFHyP3/OpenData/issues/15#issuecomment-1876052839 before we deleted and re-created `its-live-data`.

**Comment by jtherrmann** [comment]
I generated a sorted list of keys from each inventory report using @asjohnston-asf 's `sort` command:

```
cut -d '"' -f 4 *.csv | sort -T ~/sort-tmp/ --parallel=8 --buffer-size=14G --output=sorted_keys.txt
```

I confirmed that the two lists are the same:

```
$ du open-inventory/its-live-open_sorted_keys.txt data-inventory/its-live-data_sorted_keys.txt 
52165584        open-inventory/its-live-open_sorted_keys.txt
52165584        data-inventory/its-live-data_sorted_keys.txt
$ sha256sum open-inventory/its-live-open_sorted_keys.txt data-inventory/its-live-data_sorted_keys.txt 
1e546832c5a0e9dcfc871aa62dde573b5f7c7d47890b923129902ffc4bb89603  open-inventory/its-live-open_sorted_keys.txt
1e546832c5a0e9dcfc871aa62dde573b5f7c7d47890b923129902ffc4bb89603  data-inventory/its-live-data_sorted_keys.txt
```

**Comment by jtherrmann** [comment]
I think this issue can be marked done whenever we want.

**Comment by jhkennedy** [comment]
I am planning on closing it once we've deleted `s3://its-live-open`

**Comment by jhkennedy** [comment]
And all finished :tada:
